### ISSUE TITLE
E2E Testing

### ISSUE THREAD

**Issue by yorkxin** [yorkxin/copy-as-markdown#135]
I can only do so much with manual testing. Now that I started bloating features, it is time to add E2E tests before I regret.

I've studied [E2E tests on Chrome's Guide](https://developer.chrome.com/docs/extensions/how-to/test/end-to-end-testing). I've also found that there are limits on Puppeteer in terms of testing extensions:

- [\`page.keyboard.\*\`: pressing Ctrl-F doesn't act as pressing it manually · Issue #2210 · puppeteer/puppeteer](https://github.com/puppeteer/puppeteer/issues/2210)
- [尝试用 Puppeteer 测试扩展程序 · Issue #107 · lmk123/blog](https://github.com/lmk123/blog/issues/107)
- [对扩展程序进行 E2E 测试的踩坑记录 · Issue #122 · lmk123/blog](https://github.com/lmk123/blog/issues/122)

The second article suggested that Selenium is a better choice when it comes to inputs controlled by the OS (e.g. mouse right-click, keypress). Therefore it might be a better idea to start with Selenium.

**Comment by yorkxin** [comment]
When trying to test feature #143 which requires context menu click on the tabs bar, I found it really hard to achieve with AWT Robot library, because I'll need to move the mouse to a position where the tabs bar is. Moving focus by keyboard (F6 then Shift-Tab * 4) isn't enough.

Also I learned from a StackOverflow post (lost link) that Selenium is launched by another program (in my case it's something something remote TestNG), so to focus on the browser itself, it is necessary to use Cmd+Tab (on macOS).
